File: lambda-functions/update-category.js
// AWS Lambda Function - Update Category (Node.js)
// Matches your existing eventregistrations table structure
const { DynamoDBClient, ScanCommand } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocumentClient, UpdateCommand } = require("@aws-sdk/lib-dynamodb");

// Initialize DynamoDB client
const REGION = "ap-south-1";
const client = new DynamoDBClient({ region: REGION });
const docClient = DynamoDBDocumentClient.from(client);

// Your existing table name
const TABLE_NAME = 'eventregistrations';

const corsHeaders = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS'
};

exports.handler = async (event) => {
    try {
        // Handle OPTIONS for CORS preflight
        if (event.requestContext?.http?.method === 'OPTIONS' || event.httpMethod === 'OPTIONS') {
            return {
                statusCode: 200,
                headers: corsHeaders,
                body: ''
            };
        }

        // Parse request body
        const body = typeof event.body === 'string' ? JSON.parse(event.body) : event;
        const { registrationId, category, assignedCategorizer, pk, sk } = body;

        // Validate required fields
        if (!registrationId) {
            return {
                statusCode: 400,
                headers: corsHeaders,
                body: JSON.stringify({
                    success: false,
                    message: 'registrationId is required'
                })
            };
        }

        let pkValue = pk;
        let skValue = sk;

        // If pk/sk not provided, fallback to Scan (but correctly this time!)
        if (!pkValue || !skValue) {
            console.log(`⚠️ Missing pk/sk for ${registrationId}, performing scan...`);
            const scanCommand = new ScanCommand({
                TableName: TABLE_NAME,
                FilterExpression: 'registrationId = :regId',
                ExpressionAttributeValues: {
                    ':regId': { S: registrationId }
                }
                // REMOVED Limit: 1 because it breaks the scan!
            });

            const scanResult = await client.send(scanCommand);

            if (!scanResult.Items || scanResult.Items.length === 0) {
                return {
                    statusCode: 404,
                    headers: corsHeaders,
                    body: JSON.stringify({
                        success: false,
                        message: `Candidate with registrationId ${registrationId} not found`
                    })
                };
            }

            const item = scanResult.Items[0];
            pkValue = item.pk?.S;
            skValue = item.sk?.S;
        }

        if (!pkValue || !skValue) {
            return {
                statusCode: 500,
                headers: corsHeaders,
                body: JSON.stringify({
                    success: false,
                    message: `Missing pk/sk for registrationId ${registrationId}`
                })
            };
        }

        // Prepare update expression
        let updateExpression = 'SET category = :category, categorizedAt = :categorizedAt';
        const expressionValues = {
            ':category': category,
            ':categorizedAt': new Date().toISOString()
        };

        // Add categorizer info if provided
        if (assignedCategorizer) {
            updateExpression += ', assignedCategorizer = :assignedCategorizer';
            expressionValues[':assignedCategorizer'] = assignedCategorizer;
        }

        // Update the item in DynamoDB using pk and sk
        const updateCommand = new UpdateCommand({
            TableName: TABLE_NAME,
            Key: {
                pk: pkValue,
                sk: skValue
            },
            UpdateExpression: updateExpression,
            ExpressionAttributeValues: expressionValues,
            ReturnValues: 'ALL_NEW'
        });

        const response = await docClient.send(updateCommand);

        console.log(`✅ Successfully updated category for ${registrationId}: ${category}`);

        return {
            statusCode: 200,
            headers: corsHeaders,
            body: JSON.stringify({
                success: true,
                message: `Category updated successfully to ${category}`,
                data: {
                    registrationId,
                    category,
                    assignedCategorizer,
                    pk: pkValue,
                    sk: skValue
                }
            })
        };

    } catch (error) {
        console.error('❌ Error updating category:', error);
        return {
            statusCode: 500,
            headers: corsHeaders,
            body: JSON.stringify({
                success: false,
                message: `Error updating category: ${error.message}`
            })
        };
    }
};

File: lambda-functions/update-category.py
import json
import boto3
from datetime import datetime
from decimal import Decimal

# Initialize DynamoDB client
dynamodb = boto3.resource('dynamodb')

# TODO: Replace with your actual table name
CANDIDATES_TABLE_NAME = 'your-candidates-table-name'  # e.g., 'idd-candidates'
table = dynamodb.Table(CANDIDATES_TABLE_NAME)

def lambda_handler(event, context):
    """
    Lambda function to update candidate category
    
    Expected payload:
    {
        "registrationId": "REG123",
        "category": "Singing",
        "assignedCategorizer": "CAT001"
    }
    """
    
    try:
        # Parse request body
        if isinstance(event.get('body'), str):
            body = json.loads(event['body'])
        else:
            body = event
        
        # Extract parameters
        registration_id = body.get('registrationId')
        category = body.get('category')
        assigned_categorizer = body.get('assignedCategorizer')
        
        # Validate required fields
        if not registration_id:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Headers': 'Content-Type',
                    'Access-Control-Allow-Methods': 'POST, OPTIONS'
                },
                'body': json.dumps({
                    'success': False,
                    'message': 'registrationId is required'
                })
            }
        
        if not category:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Headers': 'Content-Type',
                    'Access-Control-Allow-Methods': 'POST, OPTIONS'
                },
                'body': json.dumps({
                    'success': False,
                    'message': 'category is required'
                })
            }
        
        # Prepare update expression
        update_expression = "SET category = :category, categorizedAt = :categorizedAt"
        expression_values = {
            ':category': category,
            ':categorizedAt': datetime.utcnow().isoformat()
        }
        
        # Add categorizer info if provided
        if assigned_categorizer:
            update_expression += ", assignedCategorizer = :assignedCategorizer"
            expression_values[':assignedCategorizer'] = assigned_categorizer
        
        # Update the item in DynamoDB
        # Assuming pk is the partition key - adjust based on your table structure
        response = table.update_item(
            Key={
                'pk': registration_id,  # Adjust if your key structure is different
                # 'sk': 'CANDIDATE'  # Add if you use sort key
            },
            UpdateExpression=update_expression,
            ExpressionAttributeValues=expression_values,
            ReturnValues='ALL_NEW'
        )
        
        print(f"Successfully updated category for {registration_id}: {category}")
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'POST, OPTIONS'
            },
            'body': json.dumps({
                'success': True,
                'message': f'Category updated successfully to {category}',
                'data': {
                    'registrationId': registration_id,
                    'category': category,
                    'assignedCategorizer': assigned_categorizer
                }
            }, default=str)
        }
        
    except Exception as e:
        print(f"Error updating category: {str(e)}")
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'POST, OPTIONS'
            },
            'body': json.dumps({
                'success': False,
                'message': f'Error updating category: {str(e)}'
            })
        }
